how-migration-discovery-should-work.md

```diff
@@ -0,0 +1,74 @@
+# How migration discovery should work
+
+Let's look at an example task that has had multiple versions and migrations.
+
+```text
+task/source-build
+├── 0.2
+│   ├── migrations
+│   │   │── 0.2.sh
+│   │   └── 0.2.1.sh
+│   └── source-build.yaml   <- has app.kubernetes.io/version: "0.2.1"
+└── 0.3
+    ├── migrations
+    │   │── 0.3.sh
+    │   └── 0.3.1.sh
+    └── source-build.yaml   <- has app.kubernetes.io/version: "0.3.1"
+```
+
+* We build both versions of `task/source-build`.
+* When building `task/source-build/0.2`, we look at the version label and see `0.2.1`.
+  * We attach `migrations/0.2.1.sh`
+  * We tag it with `:0.2.1` and `:0.2.1-{timestamp or git_sha or something unique}`
+* Same for `task/source-build/0.3`, but with `0.3.1`
+
+Let's say we've been doing it this way for a while and have released all the versions
+(`0.2`, `0.2.1`, `0.3`, `0.3.1`) a couple times. In the registry, the situation may
+look something like this:
+
+```text
+quay.io/konflux-ci/tekton-catalog/task-source-build
+
+  "the 0.2 branch"                          "the 0.3 branch"
+
+      0.2.1-c   <- migrations/0.2.1.sh
+        |                                       0.3.1-b     <- migrations/0.3.1.sh
+        |                                         |
+        |                                       0.3.1-a     <- migrations/0.3.1.sh
+      0.2.1-b   <- migrations/0.2.1.sh            |
+        |                                         |
+      0.2.1-a   <- migrations/0.2.1.sh            |
+        |                                        0.3-b      <- migrations/0.3.sh
+        |                                         |
+        |                                        0.3-a      <- migrations/0.3.sh
+       0.2-b    <- migrations/0.2.sh
+        |
+       0.2-a    <- migrations/0.2.sh
+```
+
+*In reality, this would be a single linear timeline of tags. It's split in
+two "branches" here for easier visualization.*
+
+When MintMaker updates a user's pipeline from `task-source-build:0.2` to
+`task-source-build:0.3.1`, we discover the migrations as follows:
+
+* List all tags for `quay.io/konflux-ci/tekton-catalog/task-source-build`
+* Sort and group them by version:
+  * `[[0.2-a, 0.2-b], [0.2.1-a, 0.2.1-b, 0.2.1-c], [0.3-a, 0.3-b], [0.3.1-a, 0.3.1-b]]`
+* For each version `>0.2, <=0.3.1`, pick an arbitrary tag from its group and download
+  the attached migration script
+* Done
+
+Why is picking an arbitrary tag okay? Because the migration script for a specific
+version is not allowed to change. We have a [CI check] for that. The CI check will
+need to be shared across all Task repos - we have [task-repo-shared-ci] for that
+(work in progress).
```

Comment:
Picking arbitrary tag may not work due to the problem mentioned in https://github.com/konflux-ci/build-definitions/pull/2611#discussion_r2309343141. PR 2611 does not address it. So, for this proposal, iteration on every tag in a group is required. However, TBH, if we implement the iteration, it looks like a technical debt.